haystack/modeling/data_handler/dataloader.py

```diff
@@ -40,19 +41,29 @@ def collate_fn(batch):
             else:
                 _tensor_names = tensor_names
 
-            if type(batch[0]) == list:
+            if isinstance(batch[0], list):
                 batch = batch[0]
 
             if len(batch[0]) != len(_tensor_names):
                 raise ModelingError(
                     f"Dataset contains {len(batch[0])} tensors while there are {len(_tensor_names)} tensor names supplied: {_tensor_names}"
                 )
-            lists_temp = [[] for _ in range(len(_tensor_names))]
-            ret = dict(zip(_tensor_names, lists_temp))
 
+            max_num_labels = self._compute_max_number_of_labels(batch=batch, tensor_names=_tensor_names)
+
+            ret = {name: [] for name in tensor_names}
             for example in batch:
                 for name, tensor in zip(_tensor_names, example):
-                    ret[name].append(tensor)
+                    padded_tensor = tensor
+
+                    # each example may have a different number of answers/labels,
+                    # so we need to pad the corresponding tensors to the max number of labels
+                    if name == "labels" and tensor.ndim > 0:
+                        num_labels = tensor.size(0)
+                        if num_labels < max_num_labels:
+                            padding = (0, 0, 0, max_num_labels - num_labels)
+                            padded_tensor = F.pad(padded_tensor, padding)
+                    ret[name].append(padded_tensor)
```

Comment:
I think we can simplify this:
```suggestion
                    # each example may have a different number of answers/labels,
                    # so we need to pad the corresponding tensors to the max number of labels
                    if name == "labels" and tensor.ndim > 0:
                        num_labels = tensor.size(0)
                        if num_labels < max_num_labels:
                            padding = (0, 0, 0, max_num_labels - num_labels)
                            tensor = F.pad(tensor, padding, value=-1)
                    ret[name].append(tensor)
```

Also, I think we should set `value=-1` in `F.pad` as this is also the value we use [here](https://github.com/deepset-ai/haystack/blob/f1932492f1bbcf18a774c3019b4f77c03045c96f/haystack/modeling/data_handler/processor.py#L628). `[0, 0]` is used for as a label for saying that the question is not answerable by the given passage.

Reply:
Very valuable suggestion!